device-server/src/main/kotlin/com/badoo/automation/deviceserver/host/AndroidEmulatorsNode.kt

```diff
@@ -0,0 +1,352 @@
+
+    override val remoteAddress: String get() = ticketHost
+
+    override fun isReachable(): Boolean {
+        // Cheapest, quietest 200-OK no-op.
+        httpGet("docker/hello").use { return it.code() == 200 }
+    }
+
+    val newEmulatorLimit = Regex("^Now: $emulatorLimit$", RegexOption.MULTILINE)
+
+    override fun prepareNode() {
+        httpPost("docker/kill-all").use { throwUnlessOK(it) }
+        httpPost("docker/set-max-emus/$emulatorLimit")
+                .use {
+                    throwUnlessOK(it)
+                    val text = it.body()?.string()
+                    if (text == null || !text.contains(newEmulatorLimit)) {
+                        throw DeviceCreationException("Failed to set emulator max: response=$it body=$text")
+                    }
+                }
+    }
+
+    override fun deleteRelease(deviceRef: DeviceRef, reason: String): Boolean {
+        val emulator = emulatorOrNullFromRef(deviceRef) ?: return false
+        val released = releaseOnTicketMachine(deviceRef)
+        val deallocated = deallocateEmulatorSlot(emulator) != null
+        refToDto.remove(deviceRef)
+        return released && deallocated
+    }
+
+    override fun getDeviceDTO(deviceRef: DeviceRef): DeviceDTO =
+            refToDto[deviceRef] ?: throw DeviceNotFoundException("No deviceDTO found for ref=$deviceRef")
+
+    override fun totalCapacity(desiredCaps: DesiredCapabilities): Int = emulatorLimit
+
+    override fun capacityRemaining(desiredCaps: DesiredCapabilities): Float = (emulatorLimit - allocatedEmulators.size.toFloat()) / emulatorLimit
+
+    override fun createDeviceAsync(desiredCaps: DesiredCapabilities): DeviceDTO {
+        val ix = serial.getAndAdd(1)
+        val ref = "${ticketHost.replace(':', '_')}__$ix".replace(Regex("[^\\w.]"), "-")
+        val model = desiredCaps.model
+        if (model == null || !model.startsWith("android-avd-")) {
+            throw IllegalArgumentException("Model must start with android-avd-, e.g. model=android-avd-oreo-snapshot")
+        }
+        val deviceDTO = updateDtoForRef(
+                ref = ref,
+                state = DeviceState.CREATING,
+                emulatorSerial = "",
+                model = model
+        )
+        concurrentBooter.submit { requestEmulatorAndAwaitBooted(ref, model) }
+        return deviceDTO
+    }
+
+    override fun list(): List<DeviceDTO> {
+        return allocatedEmulators.values.filterNotNull().mapNotNull { refToDto[it] }
+    }
+
+    private fun requestEmulatorAndAwaitBooted(ref: DeviceRef, model: String) {
+        var lastError: Throwable? = null
+        val emulator = allocateEmulatorSlot(ref)
+        try {
+            requestEmulator(ref, emulator, model)
+            awaitEmulatorOrTimeOut(ref, emulator, model)
+        } catch (e: Throwable) {
+            lastError = e
+        } finally {
+            deallocateIfNotCreated(ref, emulator, model, lastError)
```

Comment:
What is the benefit of having auto-deallocation on creation failure if the client will anyway have to call release explicitly?

Reply:
As I think we discussed, it freed up the emulator slot in case someone else wanted it. It's not a big saving - do you think it's harmful?

Reply:
[Nick indicated offline that he doesn't care]